sound/soc/sof/intel/hda-bus.c

```diff
@@ -87,22 +87,21 @@ int sof_hda_bus_init(struct hdac_bus *bus, struct device *dev,
 
 	bus->io_ops = &io_ops;
 	INIT_LIST_HEAD(&bus->stream_list);
-	INIT_LIST_HEAD(&bus->codec_list);
 
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
-	bus->ops = &bus_ops;
-	INIT_WORK(&bus->unsol_work, snd_hdac_bus_process_unsol_events);
-#endif
-	spin_lock_init(&bus->reg_lock);
-	mutex_init(&bus->cmd_mutex);
 	bus->irq = -1;
-
 	bus->ext_ops = ext_ops;
-	INIT_LIST_HEAD(&bus->hlink_list);
 	bus->idx = idx++;
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+	INIT_LIST_HEAD(&bus->codec_list);
+	INIT_LIST_HEAD(&bus->hlink_list);
+	spin_lock_init(&bus->reg_lock);
+	mutex_init(&bus->cmd_mutex);
 	mutex_init(&bus->lock);
+	bus->ops = &bus_ops;
+	INIT_WORK(&bus->unsol_work, snd_hdac_bus_process_unsol_events);
 	bus->cmd_dma_state = true;
+#endif
```

Comment:
some members such as hlink_list is a general member and will be used in other places.
Let's consume some other module such as hda-dai.c is compiled while CONFIG_SND_SOC_SOF_HDA is not enabled (Makefile and Kconfig permit this situation although I think there may not such scenario.) What if hda-dai.c uses members?

Reply:
@libinyang sorry, don't get your point. are you saying that hda-dai.c doesn't compile when CONFIG_SND_SOC_SOF_HDA is unselected?

Reply:
You are right. In this case, it is still in the CONFIG_SND_SOC_SOF_HDA.

Reply:
Not able to follow these comments, too many double-negatives. I also find suspicious that hlink_list can be parked under the SOF_HDA config, @keyonjie can you double check?

Also for the bus_idx it'd be nice to find a solution that doesn't rely on a static variable, maybe by doing the ref-counting at a higher level and passing the argument.

Reply:
I went through hlink_list with Libin and confirmed that all hlink_list related codes are embraced in #ifdef CONFIG_SND_SOC_SOF_HDA.

For bus_idx, I can imagine the potential issue here when doing module unload/reload, sorry the time is used up today, let me figure out a solution for it next Monday.

Reply:
so the list is wrong per the discussion on https://github.com/thesofproject/linux/pull/279, reg_lock is required in all cases?

Reply:
yes, let's move reg_lock out.